src/idl_gen_cpp.cpp

```diff
@@ -1118,6 +1121,39 @@ class CppGenerator : public BaseGenerator {
       code_ += "    return {{FIELD_VALUE}};";
       code_ += "  }";
 
+      if (field.value.type.base_type == BASE_TYPE_UNION) {
+        union_fields.push_back(&field);
+        auto u = field.value.type.enum_def;
+
+        code_ += "  template<typename T> "
+                "const T *{{FIELD_NAME}}_as() const;";
+
+        for (auto u_it = u->vals.vec.begin();
+             u_it != u->vals.vec.end(); ++u_it) {
+          if (!(*u_it)->struct_def) {
+            continue;
+          }
+          auto arg_struct_def = (*u_it)->struct_def;
+          auto full_struct_name = WrapInNameSpace(*arg_struct_def);
+
+          // @TODO: Mby make this decisions more universal? How?
+          code_.SetValue("U_GET_TYPE", field.name + UnionTypeFieldSuffix());
+          code_.SetValue("U_ELEMENT_TYPE", WrapInNameSpace(
+                         u->defined_namespace, GetEnumValUse(*u, **u_it)));
+          code_.SetValue("U_FIELD_TYPE", "const " + full_struct_name + " *");
+          code_.SetValue("U_ELEMENT_NAME", full_struct_name);
+          code_.SetValue("U_FIELD_NAME",
+                         field.name + "_as_" + (*u_it)->name);
+
+          // `const Type *union_name_asType() const` accessor.
+          code_ += "  {{U_FIELD_TYPE}}{{U_FIELD_NAME}}() const {";
+          code_ += "    return ({{U_GET_TYPE}}() == {{U_ELEMENT_TYPE}})? "
+                  "static_cast<{{U_FIELD_TYPE}}>({{FIELD_NAME}}()) "
```

Comment:
please add the generated code to the PR (run `sh generate_code.sh` in the tests dir.